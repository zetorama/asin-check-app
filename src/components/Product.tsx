import React, { Component } from 'react'
import { ListGroup, ListGroupItem, Button, Alert } from 'reactstrap'

import { GrabbedProduct, GrabStatus } from '../product'

const InfoItem = ({ label, children }: { label: string; children: React.ReactNode }) => (
	<ListGroupItem>
		<strong className='mr-2'>{label}:</strong>
		{children}
	</ListGroupItem>
)

export interface ProductProps {
	product: GrabbedProduct
}

export class Product extends Component<ProductProps> {
	render() {
		const { product } = this.props
		if (!product) return null

		return (
			<div className='well'>
				<h1>
					ASIN:
					<code>{product.asin}</code>
				</h1>
				{product.data && product.data.title && <h2>{product.data.title}</h2>}
				<ListGroup>
					{product.data ? this.renderDetails() : this.renderExcuse()}
					<InfoItem label='Updated On'>
						{String(product.updatedOn)}{' '}
						<Button color='primary' size='sm' outline>
							Refresh
						</Button>
					</InfoItem>
				</ListGroup>
			</div>
		)
	}

	renderExcuse() {
		const { status } = this.props.product
		switch (status) {
			case GrabStatus.Timeout:
				return (
					<ListGroupItem>
						<Alert color='warning'>Sync was canceled due to timeout. Please, try again</Alert>
					</ListGroupItem>
				)
			case GrabStatus.Nomatch:
				return (
					<ListGroupItem>
						<Alert color='warning'>Product with such ASIN wasn't found on Amazon. Sorry for that</Alert>
					</ListGroupItem>
				)
			case GrabStatus.Fail:
				return (
					<ListGroupItem>
						<Alert color='danger'>There was some fatal error during last sync. I wish I had logs</Alert>
					</ListGroupItem>
				)
		}
	}

	renderDetails() {
		const { data } = this.props.product
		if (!data) return null

		return (
			<>
				<InfoItem label='Category'>{data.categories.join(' › ')}</InfoItem>
				<InfoItem label='Best Sellers Rank'>{this.renderRank()}</InfoItem>
				<InfoItem label='Product Dimensions'>{this.renderDimensions()}</InfoItem>
			</>
		)
	}

	renderDimensions() {
		const { data } = this.props.product
		if (!data) return null

		return data.dimensions || '—'
	}

	renderRank() {
		const { data } = this.props.product
		if (!data) return null

		if (!data.rank) {
			return '—'
		}

		return data.rank.map((rank) => (
			<>
				<br />
				{rank}
			</>
		))
	}
}

export default Product
