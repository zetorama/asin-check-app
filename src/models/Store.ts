import { types, getParent, getSnapshot, applySnapshot, Instance, SnapshotIn } from 'mobx-state-tree'

import { fetchProductsLog, saveProductsLog, fetchProductsByAsins, refreshProductsByAsins } from '../services/product'
import { Product, ProductAsin, ProductModel, ProductStatus, ProductSnapshot } from './Product'

export const PRODUCTS_LIMIT = 12

export type StoreModel = Instance<typeof Store>
export type StoreSnapshot = SnapshotIn<typeof Store>

type Void<T> = { [P in keyof T]?: undefined }

const resetObj = <T>(source: T): Void<T> =>
    Object.keys(source).reduce((all, one) => Object.assign(all, { [one]: undefined }), {})

export const ProductInStore = types.compose(
    Product,
    types.model('ProductInStore').views((self) => ({
        get isCurrentlyViewing(): boolean {
            // assumming, parent has an `activeProduct` as a ref
            return getParent<StoreModel>(self, 2).viewingProduct === self
        },
    })),
)

export type ProductInStoreModel = Instance<typeof ProductInStore>
export type ProductInStoreSnapshot = SnapshotIn<typeof ProductInStore>

export const Store = types
    .model('AppStore', {
        version: '0.0.0',
        products: types.array(ProductInStore),
        viewingProduct: types.maybe(types.reference(ProductInStore)),
        lookingUpValue: types.optional(types.string, ''),
    })
    .views((self) => ({
        get loadingProducts(): ProductModel[] {
            return self.products.filter((p) => p.isLoading)
        },
        getProduct(asin: ProductAsin): ProductModel | undefined {
            return self.products.find((p) => p.asin === asin)
        },
    }))
    .actions((self) => ({
        mergeUpdates(products: ProductSnapshot[]) {
            // Update only existed items, keeping the order
            for (const update of products) {
                const product = self.getProduct(update.asin)
                if (product) {
                    applySnapshot(product, Object.assign(resetObj(product), update))

                    if (product === self.viewingProduct) {
                        // Mark as seen, as its currently opened
                        product.toggleViewed(true)
                    }
                }
            }
        },
        createProduct(asin: ProductAsin): ProductInStoreModel {
            const product = self.getProduct(asin)
            const defaults = product ? getSnapshot(product) : undefined
            return ProductInStore.create({
                ...defaults,
                asin,
                status: ProductStatus.New,
            })
        },
        addProducts(products: ProductModel[]) {
            if (!products || !Array.isArray(products) || !products.length) return

            const oldProducts = self.products.filter((p1) => !products.find((p2) => p1.asin === p2.asin))
            const newProducts: ProductInStoreModel[] = products.concat(oldProducts).slice(0, PRODUCTS_LIMIT)
            self.products.replace(newProducts)
        },
        clearLookingUpValue() {
            self.lookingUpValue = ''
        },
        viewProduct(product?: ProductModel | ProductAsin) {
            self.viewingProduct = product
                ? self.getProduct(ProductInStore.is(product) ? product.asin : product)
                : undefined
            if (self.viewingProduct && !self.viewingProduct.isViewed) {
                self.viewingProduct.toggleViewed(true)
            }
        },
    }))
    .actions((self) => ({
        saveProductsLog() {
            saveProductsLog(getSnapshot(self.products))
        },
        async fetchProductsLog() {
            const products = (await fetchProductsLog()) as ProductModel[]
            products && products.length && self.addProducts(products)
        },
        async lookup(value: string) {
            self.lookingUpValue = String(value || '')

            if (self.lookingUpValue) {
                const asins = self.lookingUpValue.split(/\W+/)
                const products = asins
                    .filter((asin, index) => asin && asins.indexOf(asin) === index)
                    .map((asin) => self.createProduct(asin))

                self.viewingProduct = undefined
                await self.addProducts(products)

                // switch to a newly added product
                self.viewProduct(self.products[0])
                self.clearLookingUpValue()
            }
        },
        async fetchUpdates(asins: ProductAsin[] = self.loadingProducts.map((p) => p.asin)) {
            const update = await fetchProductsByAsins(asins)
            update && self.mergeUpdates(update)
        },
        async refreshProducts(asins: ProductAsin[]) {
            for (const asin of asins) {
                const product = self.getProduct(asin)
                product && product.setStatus(ProductStatus.InQueue)
            }
            const update = await refreshProductsByAsins(asins)
            update && self.mergeUpdates(update)
        },
    }))

export default Store
