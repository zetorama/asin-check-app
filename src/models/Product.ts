import { getParent, types, Instance, SnapshotIn } from 'mobx-state-tree'

import { StoreModel } from './Store'

export enum ProductStatus {
    Init = 'init',
    Queue = 'queue',
    Fetch = 'fetch',
    Success = 'success',
    Nomatch = 'nomatch',
    Timeout = 'timeout',
    Fail = 'fail',
}

export const ProductDetails = types.model('ProductDetails', {
    title: types.string,
    categories: types.array(types.string),
    dimensions: types.maybe(types.string),
    rank: types.maybe(types.array(types.string)),
    rating: types.maybe(types.string),
})

export type ProductAsin = string

export type ProductModel = Instance<typeof Product>
export type ProductSnapshot = SnapshotIn<typeof Product>

export const Product = types
    .model('Product', {
        asin: types.identifier,
        status: types.enumeration<ProductStatus>('ProductStatus', Object.values(ProductStatus)),
        data: types.maybe(types.frozen(ProductDetails)),

        createdOn: types.maybe(types.string),
        updatedOn: types.maybe(types.string),
        isViewed: false,
    })
    .views((self) => ({
        get isLoading(): boolean {
            return (
                self.status === ProductStatus.Init ||
                self.status === ProductStatus.Queue ||
                self.status === ProductStatus.Fetch
            )
        },
        get isCurrentlyViewing(): boolean {
            // assumming, parent has an `activeProduct` as a ref
            return getParent<StoreModel>(self, 2).viewingProduct === self
        },
    }))
    .actions((self) => ({
        toggleViewed(setAsViewed: boolean = !self.isViewed) {
            self.isViewed = setAsViewed
        },
    }))

export default Product
