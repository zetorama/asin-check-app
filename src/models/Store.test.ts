import { getSnapshot } from 'mobx-state-tree'
import { Store } from './Store'
import { Product, ProductStatus } from './Product'

describe('Store model', () => {
    it('creates an empty model', () => {
        const model = Store.create()

        expect(getSnapshot(model)).toMatchSnapshot()
    })

    it('creates model with two products', () => {
        const model = Store.create({
            products: [{ asin: 'aaa', status: ProductStatus.Success }, { asin: 'bbb', status: ProductStatus.Fail }],
        })

        expect(model.products.length).toBe(2)

        expect(getSnapshot(model)).toMatchSnapshot()
    })

    it('gets a product', () => {
        const model = Store.create({
            products: [{ asin: 'aaa', status: ProductStatus.Success }, { asin: 'bbb', status: ProductStatus.Fail }],
        })

        expect(model.getProduct('bbb')).toHaveProperty('asin', 'bbb')
        expect(model.getProduct('zzz')).toBe(undefined)
    })

    it('creates a product', () => {
        const model = Store.create()
        const product = model.createProduct('xxx')
        expect(product).toMatchSnapshot()
    })

    it('adds a product', () => {
        const model = Store.create()

        expect(model.products.length).toBe(0)
        model.addProducts([model.createProduct('xxx')])
        expect(model.products.length).toBe(1)

        expect(getSnapshot(model)).toMatchSnapshot()
    })

    it('merges updates of only existed products', () => {
        const model = Store.create({
            products: [{ asin: 'aaa', status: ProductStatus.Success }, { asin: 'bbb', status: ProductStatus.Fail }],
        })

        expect(model.products.length).toBe(2)
        model.mergeUpdates([
            { asin: 'zzz', status: ProductStatus.Refreshing },
            { asin: 'bbb', status: ProductStatus.Nomatch },
        ])

        expect(getSnapshot(model)).toMatchSnapshot()
    })

    it('switches viewing product', () => {
        const model = Store.create({
            products: [{ asin: 'aaa', status: ProductStatus.Success }, { asin: 'bbb', status: ProductStatus.Fail }],
        })

        expect(model.viewingProduct).toBe(undefined)
        model.viewProduct('aaa')
        expect(model.viewingProduct).toHaveProperty('asin', 'aaa')
        model.viewProduct('bbb')
        expect(model.viewingProduct).toHaveProperty('asin', 'bbb')
        model.viewProduct(undefined)
        expect(model.viewingProduct).toBe(undefined)
    })
})
