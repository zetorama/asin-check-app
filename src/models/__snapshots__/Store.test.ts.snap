// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Store model adds a product 1`] = `
Object {
  "lookingUpValue": "",
  "products": Array [
    Object {
      "asin": "xxx",
      "createdOn": undefined,
      "data": undefined,
      "isViewed": false,
      "status": "new",
      "updatedOn": undefined,
    },
  ],
  "version": "0.0.0",
  "viewingProduct": undefined,
}
`;

exports[`Store model creates a product 1`] = `
Object {
  "asin": "xxx",
  "createdOn": undefined,
  "data": undefined,
  "isViewed": false,
  "status": "new",
  "updatedOn": undefined,
}
`;

exports[`Store model creates an empty model 1`] = `
Object {
  "lookingUpValue": "",
  "products": Array [],
  "version": "0.0.0",
  "viewingProduct": undefined,
}
`;

exports[`Store model creates model with two products 1`] = `
Object {
  "lookingUpValue": "",
  "products": Array [
    Object {
      "asin": "aaa",
      "createdOn": undefined,
      "data": undefined,
      "isViewed": false,
      "status": "success",
      "updatedOn": undefined,
    },
    Object {
      "asin": "bbb",
      "createdOn": undefined,
      "data": undefined,
      "isViewed": false,
      "status": "fail",
      "updatedOn": undefined,
    },
  ],
  "version": "0.0.0",
  "viewingProduct": undefined,
}
`;

exports[`Store model merges updates of only existed products 1`] = `
Object {
  "lookingUpValue": "",
  "products": Array [
    Object {
      "asin": "aaa",
      "createdOn": undefined,
      "data": undefined,
      "isViewed": false,
      "status": "success",
      "updatedOn": undefined,
    },
    Object {
      "asin": "bbb",
      "createdOn": undefined,
      "data": undefined,
      "isViewed": false,
      "status": "nomatch",
      "updatedOn": undefined,
    },
  ],
  "version": "0.0.0",
  "viewingProduct": undefined,
}
`;
